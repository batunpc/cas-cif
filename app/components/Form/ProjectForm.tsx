import type { JSONSchema7 } from "json-schema";
import FormBase from "../Form/FormBase";
import FormComponentProps from "../Form/FormComponentProps";
import { graphql, useFragment } from "react-relay";
import type { ProjectForm_query$key } from "__generated__/ProjectForm_query.graphql";
import { useMemo } from "react";
import SelectRfpWidget from "components/Form/SelectRfpWidget";
import SelectProjectStatusWidget from "./SelectProjectStatusWidget";

interface Props extends FormComponentProps {
  query: ProjectForm_query$key;
}

const ProjectForm: React.FC<Props> = (props) => {
  const { query } = useFragment(
    graphql`
      fragment ProjectForm_query on Query {
        query {
          allOperators {
            edges {
              node {
                rowId
                legalName
                tradeName
                bcRegistryId
              }
            }
          }
          ...SelectRfpWidget_query
          ...SelectProjectStatusWidget_query
        }
      }
    `,
    props.query
  );

  let selectedOperator = useMemo(() => {
    return query.allOperators.edges.find(
      ({ node }) => node.rowId === props.formData.operatorId
    );
  }, [query, props.formData.operatorId]);

  const schema: JSONSchema7 = useMemo(() => {
    return {
      type: "object",
      required: [
        "rfpNumber",
        "projectName",
        "summary",
        "operatorId",
        "fundingStreamRfpId",
      ],
      properties: {
        rfpNumber: {
          type: "string",
          title: "RFP Number",
          pattern: "^((\\d{4})-RFP-([1-2])-(\\d{3,4})-([A-Z]{4}))$",
        },
        projectName: { type: "string", title: "Project Name" },
        summary: { type: "string", title: "Summary" },
        operatorId: {
          type: "number",
          title: "Legal Operator Name and BC Registry ID",
          default: undefined,
          anyOf: query.allOperators.edges.map(({ node }) => {
            return {
              type: "number",
              title: `${node.legalName} (${node.bcRegistryId})`,
              enum: [node.rowId],
              value: node.rowId,
            };
          }),
        },
        operatorTradeName: {
          type: "string",
        },
        fundingStreamRfpId: {
          type: "number",
          title: "Funding Stream RFP ID",
          default: undefined,
        },
        projectStatusId: {
          type: "number",
          title: "Project Status",
          default: undefined,
        },
      },
    };
  }, [query]);

  const uiSchema = useMemo(() => {
    return {
      rfpNumber: {
        "ui:placeholder": "2020-RFP-1-456-ABCD",
        "ui:col-md": 12,
        "bcgov:size": "small",
      },
      projectName: {
        "ui:placeholder": "Short project name",
        "ui:col-md": 12,
        "bcgov:size": "small",
      },
      summary: {
        "ui:col-md": 12,
        "ui:widget": "TextAreaWidget",
        "bcgov:size": "small",
      },
      operatorId: {
        "ui:placeholder": "Select an Operator",
        "ui:col-md": 12,
        "bcgov:size": "small",
        "ui:widget": "SearchWidget",
      },
      operatorTradeName: {
        "ui:col-md": 12,
        "ui:widget": "DisplayOnly",
        "bcgov:size": "small",
        "ui:options": {
          text: `${selectedOperator ? selectedOperator.node.tradeName : ""}`,
          title: "Trade Name",
        },
      },
      fundingStreamRfpId: {
        "ui:widget": "SelectRfpWidget",
        "ui:col-md": 12,
        "bcgov:size": "small",
      },
      projectStatusId: {
        "ui:placeholder": "Select a Project Status",
        "ui:widget": "SelectProjectStatusWidget",
        "ui:col-md": 12,
        "bcgov:size": "small",
      },
    };
  }, [selectedOperator]);

  return (
    <FormBase
      {...props}
      schema={schema}
      uiSchema={uiSchema}
      formContext={{ query, form: props.formData }}
      widgets={{
        SelectRfpWidget: SelectRfpWidget,
        SelectProjectStatusWidget: SelectProjectStatusWidget,
      }}
    />
  );
};

export default ProjectForm;
