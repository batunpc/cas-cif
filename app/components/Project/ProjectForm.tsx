import type { JSONSchema7 } from "json-schema";
import FormBase from "../Form/FormBase";
import FormComponentProps from "../Form/FormComponentProps";
import { graphql, useFragment } from "react-relay";
import type { ProjectForm_query$key } from "__generated__/ProjectForm_query.graphql";
import { useMemo } from "react";

interface Props extends FormComponentProps {
  query: ProjectForm_query$key;
}

const ProjectForm: React.FC<Props> = (props) => {
  const { query } = useFragment(
    graphql`
      fragment ProjectForm_query on Query {
        query {
          allOperators {
            edges {
              node {
                rowId
                legalName
                tradeName
                bcRegistryId
              }
            }
          }
          allFundingStreams {
            edges {
              node {
                rowId
                name
                description
              }
            }
          }
        }
      }
    `,
    props.query
  );

  let selectedOperator = useMemo(() => {
      return query.allOperators.edges.find(
        ({ node }) => node.rowId === props.formData.operatorId
      )
    }, [query, props.formData.operatorId]);

  let selectedFundingStream = useMemo(() => {
      return query.allFundingStreams.edges.find(
        ({ node }) => node.rowId === props.formData.fundingStreamId
      )
    }, [query, props.formData.fundingStreamId]);

  const schema: JSONSchema7 = useMemo(() => {
    return {
      type: "object",
      required: ["rfpNumber", "description", "operatorId", "fundingStreamId"],
      properties: {
        rfpNumber: {
          type: "string",
          title: "RFP Number",
          pattern: "^((\\d{4})-RFP-([1-2])-(\\d{3,4})-([A-Z]{4}))$",
        },
        description: { type: "string", title: "Description" },
        operatorId: {
          type: "number",
          title: "Legal Operator Name and BC Registry ID",
          default: undefined,
          anyOf: query.allOperators.edges.map(({ node }) => {
            return {
              type: "number",
              title: `${node.legalName} (${node.bcRegistryId})`,
              enum: [node.rowId],
              value: node.rowId,
            };
          }),
        },
        operatorTradeName: {
          type: "string",
        },
        fundingStreamId: {
          type: "number",
          title: "Funding Stream ID",
          default: undefined,
          anyOf: query.allFundingStreams.edges.map(({ node }) => {
            return {
              type: "number",
              title: node.name,
              enum: [node.rowId],
              value: node.rowId,
            };
          }),
        },
        fundingStreamDescription: {
          type: "string",
        },
      },
    };
  }, [query]);

  const uiSchema = useMemo(() => {
    return {
      rfp_number: {
        "ui:placeholder": "2020-RFP-1-456-ABCD",
        "ui:col-md": 12,
        "bcgov:size": "small",
      },
      description: {
        "ui:placeholder": "describe the project...",
        "ui:col-md": 12,
        "bcgov:size": "small",
      },
      operatorId: {
        "ui:placeholder": "Select an Operator",
        "ui:col-md": 12,
        "bcgov:size": "small",
        "ui:widget": "SearchWidget",
      },
      operatorTradeName: {
        "ui:col-md": 12,
        "ui:widget": "DisplayOnly",
        "bcgov:size": "small",
        "ui:options": {
          text: `${selectedOperator ? selectedOperator.node.tradeName : ""}`,
          title: "Trade Name",
        },
      },
      fundingStreamId: {
        "ui:placeholder": "Select a Funding Stream",
        "ui:widget": "select",
        "ui:col-md": 12,
        "bcgov:size": "small",
      },
      fundingStreamDescription: {
        "ui:col-md": 12,
        "ui:widget": "DisplayOnly",
        "bcgov:size": "large",
        "ui:options": {
          text: `${
            selectedFundingStream ? selectedFundingStream.node.description : ""
          }`,
          title: "Description",
        },
      },
    };
  }, [selectedOperator, selectedFundingStream]);

  return <FormBase {...props} schema={schema} uiSchema={uiSchema} />;
};

export default ProjectForm;
