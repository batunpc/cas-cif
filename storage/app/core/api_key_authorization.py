from fastapi import Depends, Header
from starlette.exceptions import HTTPException
from starlette.status import HTTP_403_FORBIDDEN
from fastapi.security import APIKeyHeader
from app.schema import ApiKey
from app.config import STORAGE_API_KEY


def _get_api_key(api_key: str = Header(...)) -> str:
    return api_key


def get_api_key_authorizer(*, required: bool = True) -> ApiKey:
    return _get_authorize_api_request


async def _get_authorize_api_request(api_key: str = Depends(_get_api_key)) -> ApiKey:
    """
    checks whether the api key matches the environment variable
    """
    try:
        if api_key == STORAGE_API_KEY:
            return ApiKey(api_key=api_key)
        else:
            raise Exception('invalid api key')
    except Exception as exc:
        raise HTTPException(
            status_code=HTTP_403_FORBIDDEN, detail="invalid api key"
        )

auth_scheme = APIKeyHeader
