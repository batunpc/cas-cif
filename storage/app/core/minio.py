import logging
from os import path
from minio import Minio
from fastapi import HTTPException
from app.config import MINIO_ACCESS_KEY, MINIO_SECRET_KEY, MINIO_HOST_URL
from starlette.status import HTTP_500_INTERNAL_SERVER_ERROR

logger = logging.getLogger("attachments")
_minio_client = None


def get_minio_client():
    if _minio_client is None:
        _minio_client = Minio(MINIO_HOST_URL,
                              access_key=MINIO_ACCESS_KEY,
                              secret_key=MINIO_SECRET_KEY,
                              secure=False)

        try:
            _minio_client.list_buckets()
            logger.info(
                f'Successfully connected to MINIO server on {MINIO_HOST_URL}')
        except Exception as exc:
            logger.error(
                "Error connecting to MINIO server on {MINIO_HOST_URL}: " + str(exc))

    return _minio_client


def s3_upload_raw_file(bucket_name: str, file_name: str, data: object, length: int):
    """
    Uploads a raw file to s3 Minio storage and returns the file response object
    """
    try:
        result = get_minio_client().put_object(bucket_name=bucket_name,
                                               object_name=file_name,
                                               data=data,
                                               length=length)
        return result
    except Exception as exc:
        raise HTTPException(status_code=HTTP_500_INTERNAL_SERVER_ERROR,
                            detail=f'Unknown Error During Upload Process {exc}')


def s3_upload_file(destination_file_name: str, local_file_path: str, content_type: str, bucket_name: str):
    """
    Uploads a file to s3 Minio storage and returns the file response object
    """
    if not path.exists(local_file_path):
        raise FileNotFoundError(
            f'file_path: {local_file_path} is not a valid file')

    try:
        result = get_minio_client().fput_object(bucket_name=bucket_name,
                                                object_name=destination_file_name,
                                                file_path=local_file_path,
                                                content_type=content_type)
        return result
    except Exception as exc:
        raise HTTPException(status_code=HTTP_500_INTERNAL_SERVER_ERROR,
                            detail=f'Unknown Error During Upload Process {exc}')


def s3_delete_file(bucket_name: str, object_name: str):
    """
    Deletes a file from s3 minio storage
    """
    try:
        print('deleting object:', object_name)
        result = get_minio_client().remove_object(bucket_name, object_name)
        return result
    except Exception as exc:
        print('error deleting file: ', exc)
        raise HTTPException(status_code=HTTP_500_INTERNAL_SERVER_ERROR,
                            detail=f'Unknown Error During Delete Process {exc}')


def s3_list_directory(bucket_name: str, folder_name: str):
    """
    Lists all files in a s3 minio storage
    """
    try:
        print('listing files for directory:', folder_name)
        files = get_minio_client().list_objects(
            bucket_name, prefix=folder_name, recursive=True)
        return files
    except Exception as exc:
        print('error listing directory:', exc)
        raise HTTPException(status_code=HTTP_500_INTERNAL_SERVER_ERROR,
                            detail=f'Unknown Error Listing Directory {exc}')


def s3_get_file_stats(bucket_name: str, object_name: str):
    """
    Gets the file stats object from s3 minio storage
    """
    try:
        stats = get_minio_client().stat_object(bucket_name, object_name)
        return stats
    except Exception as exc:
        print('error getting file stats: ', exc)
        raise HTTPException(status_code=HTTP_500_INTERNAL_SERVER_ERROR,
                            detail=f'Unknown Error Getting File Stats {exc}')


def s3_get_file(bucket_name: str, object_name: str):
    """
    Gets a file from s3 minio storage
    """
    try:
        result = get_minio_client().get_object(bucket_name, object_name)
        return result
    except Exception as exc:
        print('error getting file: ', exc)
        raise HTTPException(status_code=HTTP_500_INTERNAL_SERVER_ERROR,
                            detail=f'Unknown Error Getting File {exc}')
