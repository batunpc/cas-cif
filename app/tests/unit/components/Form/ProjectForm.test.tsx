import { render, screen, fireEvent } from "@testing-library/react";
import ProjectForm from "components/Form/ProjectForm";
import { createMockEnvironment, MockPayloadGenerator } from "relay-test-utils";
import {
  RelayEnvironmentProvider,
  useLazyLoadQuery,
  graphql,
} from "react-relay";
import compiledProjectFormQuery, {
  ProjectFormQuery,
} from "__generated__/ProjectFormQuery.graphql";
import { FormProps } from "@rjsf/core";

const loadedQuery = graphql`
  query ProjectFormQuery @relay_test_operation {
    query {
      # Spread the fragment you want to test here
      ...ProjectForm_query
    }
  }
`;

const props: FormProps<any> = {
  formData: {},
  onChange: jest.fn(),
  schema: {},
};

let environment;
const TestRenderer = () => {
  const data = useLazyLoadQuery<ProjectFormQuery>(loadedQuery, {});
  return <ProjectForm {...props} query={data.query} />;
};
const renderProjectForm = () => {
  return render(
    <RelayEnvironmentProvider environment={environment}>
      <TestRenderer />
    </RelayEnvironmentProvider>
  );
};

describe("The Project Form", () => {
  beforeEach(() => {
    environment = createMockEnvironment();
  });
  it("matches the snapshot", () => {
    environment.mock.queueOperationResolver((operation) =>
      MockPayloadGenerator.generate(operation)
    );

    environment.mock.queuePendingOperation(compiledProjectFormQuery, {});

    const componentUnderTest = renderProjectForm();
    expect(componentUnderTest.container).toMatchSnapshot();
  });
  it("triggers the applyFormChange with the proper data", () => {
    const changeSpy = jest.fn();

    props.onChange = changeSpy;
    environment.mock.queueOperationResolver((operation) =>
      MockPayloadGenerator.generate(operation)
    );

    environment.mock.queuePendingOperation(compiledProjectFormQuery, {});

    renderProjectForm();
    fireEvent.change(screen.getByLabelText("RFP Number"), {
      target: { value: "testidentifier" },
    });

    expect(changeSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        formData: {
          rfpNumber: "testidentifier",
          summary: undefined,
        },
      })
    );
    changeSpy.mockClear();

    fireEvent.change(screen.getByLabelText(/summary/i), {
      target: { value: "testsummary" },
    });

    expect(changeSpy).toHaveBeenCalledWith(
      expect.objectContaining({
        formData: {
          rfpNumber: "testidentifier",
          summary: "testsummary",
        },
      })
    );
  });
  it("loads with the correct initial form data", () => {
    props.formData = {
      rfpNumber: "12345678",
      summary: "d",
      operatorId: 1,
      fundingStreamRfpId: 1,
    };
    environment.mock.queueOperationResolver((operation) =>
      MockPayloadGenerator.generate(operation, {
        Query() {
          return {
            allOperators: {
              edges: [
                {
                  node: {
                    rowId: 1,
                    legalName: "test operator",
                    bcRegistryId: "1234abcd",
                  },
                },
              ],
            },
            allFundingStreams: {
              edges: [
                {
                  node: {
                    rowId: 1,
                    name: "EP",
                    description: "Emissions Performance",
                  },
                },
              ],
            },
          };
        },
      })
    );

    environment.mock.queuePendingOperation(compiledProjectFormQuery, {});

    renderProjectForm();

    expect(screen.getByLabelText<HTMLInputElement>("RFP Number").value).toBe(
      "12345678"
    );
    expect(screen.getByLabelText<HTMLInputElement>("Summary").value).toBe("d");
    expect(
      screen.getByPlaceholderText<HTMLSelectElement>("Select an Operator").value
    ).toBe("test operator (1234abcd)");
  });
});
