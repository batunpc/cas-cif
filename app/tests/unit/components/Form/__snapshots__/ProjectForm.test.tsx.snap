// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Project Form matches the snapshot 1`] = `
<div>
  <form
    class="rjsf"
    novalidate=""
  >
    <div
      class="sc-jRQBWg cDwDrH pg-grid-container"
      cols="10"
    >
      <div
        class="sc-gKclnd sc-iCfMLu YZhSi fQKovI pg-grid-row"
      >
        <div
          class="sc-furwcr sc-pVTFL XpgcI eqAbko pg-grid-col"
          id="test"
          width="100"
        >
          <fieldset
            class="jsx-3408663875"
            id="formborder"
          >
            <legend
              class="jsx-3408663875"
            >
              Background
            </legend>
            <div
              class="sc-crHmcD bKATQK pg-input"
            >
              <label
                class="sc-egiyK ewKLoj pg-input-label"
                for="root_rfpNumber"
              >
                RFP Number*
              </label>
              <input
                aria-label="RFP Number*"
                class="sc-bqiRlB hrAFNs pg-input-input"
                id="root_rfpNumber"
                name="root_rfpNumber-input"
                placeholder="2020-RFP-1-456-ABCD"
                required=""
                value=""
              />
            </div>
            <div
              class="sc-crHmcD bKATQK pg-input"
            >
              <label
                class="sc-egiyK ewKLoj pg-input-label"
                for="root_projectName"
              >
                Project Name*
              </label>
              <input
                aria-label="Project Name*"
                class="sc-bqiRlB hrAFNs pg-input-input"
                id="root_projectName"
                name="root_projectName-input"
                placeholder="Short project name"
                required=""
                value=""
              />
            </div>
            <div
              class="sc-crHmcD bKATQK pg-input"
            >
              <label
                class="sc-egiyK ewKLoj pg-input-label"
                for="root_description"
              >
                Description*
              </label>
              <input
                aria-label="Description*"
                class="sc-bqiRlB hrAFNs pg-input-input"
                id="root_description"
                name="root_description-input"
                placeholder="describe the project..."
                required=""
                value=""
              />
            </div>
            <label
              for="search-dropdown-root_operatorId"
            >
              Legal Operator Name and BC Registry ID*
            </label>
            <div
              aria-expanded="false"
              class="MuiAutocomplete-root MuiAutocomplete-hasClearIcon MuiAutocomplete-hasPopupIcon css-symcig-MuiAutocomplete-root"
              role="combobox"
            >
              <div
                class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
              >
                <div
                  class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-autocomplete="list"
                    aria-invalid="false"
                    autocapitalize="none"
                    autocomplete="off"
                    class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                    id="search-dropdown-root_operatorId"
                    placeholder="Select an Operator"
                    spellcheck="false"
                    type="text"
                    value=""
                  />
                  <div
                    class="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
                  >
                    <button
                      aria-label="Clear"
                      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-clearIndicator css-1glvl0p-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-clearIndicator"
                      tabindex="-1"
                      title="Clear"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                        data-testid="CloseIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                    <button
                      aria-label="Open"
                      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium MuiAutocomplete-popupIndicator css-qzbt6i-MuiButtonBase-root-MuiIconButton-root-MuiAutocomplete-popupIndicator"
                      tabindex="-1"
                      title="Open"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="ArrowDropDownIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M7 10l5 5 5-5z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-nnbavb"
                    >
                      <span
                        class="notranslate"
                      >
                        â€‹
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <span
              class="paragraph-text"
            >
              <label>
                Trade Name
              </label>
              <p />
            </span>
            <label
              class="jsx-3703916905"
              for="funding-stream-dropdown-root_fundingStreamId"
            >
              Funding Stream ID*
            </label>
            <div
              class="sc-bkkeKt iEOAMi pg-select"
            >
              <div
                class="sc-dJjYzT juyIGG pg-select-wrapper"
              >
                <select
                  aria-label="root_fundingStreamId-select"
                  class="sc-hGPBjI hXpBuL pg-select-input"
                  id="root_fundingStreamId"
                  name="root_fundingStreamId-select"
                  placeholder="Select a Funding Stream"
                  required=""
                >
                  <option
                    class="jsx-3703916905"
                  >
                    Select a Funding Stream
                  </option>
                  <option
                    class="jsx-3703916905"
                    value="42"
                  >
                    &lt;mock-value-for-field-"name"&gt;
                  </option>
                </select>
              </div>
            </div>
            <span
              class="paragraph-text"
            >
              <label>
                Description
              </label>
              <p />
            </span>
          </fieldset>
        </div>
      </div>
    </div>
  </form>
</div>
`;
